File: src/assignments/week_3/machine-learning-ex2/ex2/costFunction.m
function [J, grad] = costFunction(theta, X, y)
%COSTFUNCTION Compute cost and gradient for logistic regression
%   J = COSTFUNCTION(theta, X, y) computes the cost of using theta as the
%   parameter for logistic regression and the gradient of the cost
%   w.r.t. to the parameters.

% Initialize some useful values
m = length(y); % number of training examples

prediction = sigmoid(X * theta);

cost_y_1 = (1 - y) .* log(1 - prediction);
cost_y_0 = -1 .* y .* log(prediction);

J = 1/m * sum(cost_y_0 - cost_y_1);
grad = 1/m .* X' * (prediction - y);



% ====================== YOUR CODE HERE ======================
% Instructions: Compute the cost of a particular choice of theta.
%               You should set J to the cost.
%               Compute the partial derivatives and set grad to the partial
%               derivatives of the cost w.r.t. each parameter in theta
%
% Note: grad should have the same dimensions as theta
%








% =============================================================

end

File: src/assignments/python/week_3/machine-learning-ex2/ex2/costFunction.py
import numpy as np
from sigmoid import sigmoid

def costFunction(theta, X, y, reg_lambda = None, flattenResult=False):
    """Compute cost and gradient for logistic regression
    J = COSTFUNCTION(theta, X, y) computes the cost of using theta as the  
    parameter for logistic regression and the gradient of the cost
    w.r.t. to the parameters."""
    
    m,n = X.shape
    theta = theta.reshape((n,1))
    prediction = sigmoid(np.dot(X, theta)) #make predictions 

    cost_y_1 = np.multiply(np.subtract(1,y), np.log(1 - prediction))
    cost_y_0 = np.multiply(np.multiply(-1, y), np.log(prediction))
    
    J = (1.0/m) * np.sum(np.subtract(cost_y_0, cost_y_1))
    
    if reg_lambda:
        J =  J + (reg_lambda/(2 * m)) * np.sum(np.power(theta[1:], 2)) #don't regularize theta[0, :]
   
    return J